File: src/adoClient.ts
import * as vscode from 'vscode';
import { AuthService } from './authService';
import { LoggingService } from './LoggingService';

export class AdoUnauthorizedError extends Error {
	constructor(public readonly status: number, message: string) {
		super(message);
		this.name = 'AdoUnauthorizedError';
	}
}

export interface AdoProfile {
	id: string;
	displayName: string;
}

export interface AdoOrganization {
	accountId: string;
	accountName: string;
	accountUri: string;
}

export interface AdoProject {
	id: string;
	name: string;
	description?: string;
	url: string;
	state: string;
}

export interface AdoPipeline {
	id: number;
	name: string;
	folder: string;
	url: string;
	revision?: number;
}

export interface AdoPipelineDetail extends AdoPipeline {
	configuration?: {
		type?: string;
		path?: string;
		repository?: {
			id?: string;
			type?: string;
			fullName?: string;
			name?: string;
		};
	};
}

export interface AdoRepository {
	id: string;
	name: string;
	url?: string;
	webUrl?: string;
	defaultBranch?: string;
	project?: { id: string; name: string };
}

interface AdoListResponse<T> {
	count: number;
	value: T[];
}

interface AccountsResponse {
	count: number;
	value: Array<{ accountId: string; accountName: string; accountUri?: string }>;
}

/**
 * Thin REST wrapper around the Azure DevOps APIs we need.
 * Uses the AuthService to attach the right Authorization header
 * (Bearer for Microsoft sessions, Basic for PAT sessions).
 */
export class AdoClient {
	constructor(
		private readonly auth: AuthService,
		private readonly logger: LoggingService,
	) { }

	async getProfile(): Promise<AdoProfile> {
		const url = 'https://app.vssps.visualstudio.com/_apis/profile/profiles/me?api-version=7.1';
		return this.getJson<AdoProfile>(url);
	}

	async listOrganizations(memberId: string): Promise<AdoOrganization[]> {
		const url = `https://app.vssps.visualstudio.com/_apis/accounts?api-version=7.1&memberId=${encodeURIComponent(memberId)}`;
		const res = await this.getJson<AccountsResponse>(url);
		return res.value.map(a => ({
			accountId: a.accountId,
			accountName: a.accountName,
			accountUri: a.accountUri ?? `https://dev.azure.com/${a.accountName}`,
		}));
	}

	async listProjects(organizationName: string): Promise<AdoProject[]> {
		const url = `https://dev.azure.com/${encodeURIComponent(organizationName)}/_apis/projects?api-version=7.1&stateFilter=wellFormed&$top=1000`;
		const res = await this.getJson<AdoListResponse<AdoProject>>(url);
		return res.value;
	}

	async listPipelines(organizationName: string, projectName: string): Promise<AdoPipeline[]> {
		const url = `https://dev.azure.com/${encodeURIComponent(organizationName)}/${encodeURIComponent(projectName)}/_apis/pipelines?api-version=7.1&$top=1000`;
		const res = await this.getJson<AdoListResponse<AdoPipeline>>(url);
		return res.value;
	}

	async getPipeline(organizationName: string, projectName: string, pipelineId: number): Promise<AdoPipelineDetail> {
		const url = `https://dev.azure.com/${encodeURIComponent(organizationName)}/${encodeURIComponent(projectName)}/_apis/pipelines/${pipelineId}?api-version=7.1`;
		return this.getJson<AdoPipelineDetail>(url);
	}

	/** Look up a Git repository by id. Returns undefined on 404. */
	async getRepository(
		organizationName: string,
		projectName: string,
		repositoryId: string,
	): Promise<AdoRepository | undefined> {
		const url = `https://dev.azure.com/${encodeURIComponent(organizationName)}/${encodeURIComponent(projectName)}/_apis/git/repositories/${encodeURIComponent(repositoryId)}?api-version=7.1`;
		try {
			return await this.getJson<AdoRepository>(url);
		} catch (err) {
			if (err instanceof Error && /\b404\b/.test(err.message)) {
				return undefined;
			}
			throw err;
		}
	}

	/**
	 * Fetch the raw text content of a file from a Git repository hosted in Azure DevOps.
	 * Returns undefined if the file is missing (404) or the repo is not a TfsGit repo.
	 * When `branch` is provided, the file is fetched from that branch; otherwise the
	 * repository's default branch is used.
	 */
	async getFileContent(
		organizationName: string,
		projectName: string,
		repositoryId: string,
		path: string,
		branch?: string,
	): Promise<string | undefined> {
		const normalized = path.startsWith('/') ? path : `/${path}`;
		let url = `https://dev.azure.com/${encodeURIComponent(organizationName)}/${encodeURIComponent(projectName)}/_apis/git/repositories/${encodeURIComponent(repositoryId)}/items?path=${encodeURIComponent(normalized)}&api-version=7.1&includeContent=true&$format=text`;
		if (branch) {
			url += `&versionDescriptor.versionType=branch&versionDescriptor.version=${encodeURIComponent(branch)}`;
		}
		return this.getText(url);
	}

	/**
	 * List heads (branches) of a Git repository. Returns short branch names
	 * (without the `refs/heads/` prefix).
	 */
	async listBranches(
		organizationName: string,
		projectName: string,
		repositoryId: string,
	): Promise<string[]> {
		const url = `https://dev.azure.com/${encodeURIComponent(organizationName)}/${encodeURIComponent(projectName)}/_apis/git/repositories/${encodeURIComponent(repositoryId)}/refs?filter=heads/&api-version=7.1&$top=1000`;
		const res = await this.getJson<AdoListResponse<{ name: string }>>(url);
		return res.value
			.map(r => r.name.replace(/^refs\/heads\//, ''))
			.sort((a, b) => a.localeCompare(b));
	}

	private async getText(url: string): Promise<string | undefined> {
		const headers = this.auth.getHeaders();
		this.logger.logDebug(`GET ${url}`);
		const response = await fetch(url, { headers: headers as unknown as Record<string, string> });
		if (response.status === 404) {
			return undefined;
		}
		if (!response.ok) {
			const body = await safeReadBody(response);
			const msg = `ADO REST call failed: ${response.status} ${response.statusText} for ${url}${body ? ` :: ${body}` : ''}`;
			this.logger.logError(msg);
			if (response.status === 401 || response.status === 403) {
				throw new AdoUnauthorizedError(
					response.status,
					vscode.l10n.t('Azure DevOps rejected the credentials ({0}). The stored token may be expired or revoked.', response.status),
				);
			}
			throw new Error(msg);
		}
		return await response.text();
	}

	private async getJson<T>(url: string): Promise<T> {
		const headers = this.auth.getHeaders();
		this.logger.logDebug(`GET ${url}`);
		const response = await fetch(url, { headers: headers as unknown as Record<string, string> });

		if (!response.ok) {
			const body = await safeReadBody(response);
			const msg = `ADO REST call failed: ${response.status} ${response.statusText} for ${url}${body ? ` :: ${body}` : ''}`;
			this.logger.logError(msg);
			if (response.status === 401 || response.status === 403) {
				throw new AdoUnauthorizedError(
					response.status,
					vscode.l10n.t('Azure DevOps rejected the credentials ({0}). The stored token may be expired or revoked.', response.status),
				);
			}
			throw new Error(msg);
		}

		// ADO returns HTML on redirect-to-login when auth fails silently. Detect it.
		const contentType = response.headers.get('content-type') ?? '';
		if (!contentType.includes('application/json')) {
			throw new AdoUnauthorizedError(
				401,
				vscode.l10n.t('Unexpected non-JSON response from {0}. Authentication may have expired.', url),
			);
		}

		return await response.json() as T;
	}
}

async function safeReadBody(response: Response): Promise<string | undefined> {
	try {
		const text = await response.text();
		return text.length > 500 ? `${text.slice(0, 500)}…` : text;
	} catch {
		return undefined;
	}
}

File: src/vs2026/AzureDevOps/AdoClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using PipelinesExplorer.VisualStudio.Services;

namespace PipelinesExplorer.VisualStudio.AzureDevOps;

/// <summary>
/// Thin REST wrapper around the Azure DevOps APIs we need. Mirrors
/// <c>AdoClient</c> in the VS Code client: same endpoints, same error
/// semantics, same sort order for <see cref="ListBranchesAsync"/>.
/// </summary>
public sealed class AdoClient : IDisposable
{
    private const string ApiVersion = "7.1";

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    private readonly HttpClient _http;
    private readonly bool _ownsHttpClient;
    private readonly IAdoAuthHeaderProvider? _authProvider;
    private readonly LoggingService _logger;

    public AdoClient(LoggingService logger, IAdoAuthHeaderProvider? authProvider = null, HttpClient? httpClient = null)
    {
        _logger = logger;
        _authProvider = authProvider;
        _http = httpClient ?? new HttpClient();
        _ownsHttpClient = httpClient is null;
    }

    public void Dispose()
    {
        if (_ownsHttpClient)
        {
            _http.Dispose();
        }
    }

    public Task<AdoProfile> GetProfileAsync(CancellationToken cancellationToken = default)
    {
        var url = $"https://app.vssps.visualstudio.com/_apis/profile/profiles/me?api-version={ApiVersion}";
        return GetJsonAsync<AdoProfile>(url, cancellationToken);
    }

    public async Task<IReadOnlyList<AdoOrganization>> ListOrganizationsAsync(string memberId, CancellationToken cancellationToken = default)
    {
        var url = $"https://app.vssps.visualstudio.com/_apis/accounts?api-version={ApiVersion}&memberId={Uri.EscapeDataString(memberId)}";
        var res = await GetJsonAsync<AdoListResponse<AccountsResponseEntry>>(url, cancellationToken).ConfigureAwait(false);
        return res.Value.Select(a => new AdoOrganization
        {
            AccountId = a.AccountId,
            AccountName = a.AccountName,
            AccountUri = string.IsNullOrEmpty(a.AccountUri) ? $"https://dev.azure.com/{a.AccountName}" : a.AccountUri!,
        }).ToList();
    }

    public async Task<IReadOnlyList<AdoProject>> ListProjectsAsync(string organizationName, CancellationToken cancellationToken = default)
    {
        var url = $"https://dev.azure.com/{Uri.EscapeDataString(organizationName)}/_apis/projects?api-version={ApiVersion}&stateFilter=wellFormed&$top=1000";
        var res = await GetJsonAsync<AdoListResponse<AdoProject>>(url, cancellationToken).ConfigureAwait(false);
        return res.Value;
    }

    public async Task<IReadOnlyList<AdoPipeline>> ListPipelinesAsync(string organizationName, string projectName, CancellationToken cancellationToken = default)
    {
        var url = $"https://dev.azure.com/{Uri.EscapeDataString(organizationName)}/{Uri.EscapeDataString(projectName)}/_apis/pipelines?api-version={ApiVersion}&$top=1000";
        var res = await GetJsonAsync<AdoListResponse<AdoPipeline>>(url, cancellationToken).ConfigureAwait(false);
        return res.Value;
    }

    public Task<AdoPipelineDetail> GetPipelineAsync(string organizationName, string projectName, int pipelineId, CancellationToken cancellationToken = default)
    {
        var url = $"https://dev.azure.com/{Uri.EscapeDataString(organizationName)}/{Uri.EscapeDataString(projectName)}/_apis/pipelines/{pipelineId}?api-version={ApiVersion}";
        return GetJsonAsync<AdoPipelineDetail>(url, cancellationToken);
    }

    /// <summary>Look up a Git repository by id. Returns <c>null</c> on 404.</summary>
    public async Task<AdoRepository?> GetRepositoryAsync(string organizationName, string projectName, string repositoryId, CancellationToken cancellationToken = default)
    {
        var url = $"https://dev.azure.com/{Uri.EscapeDataString(organizationName)}/{Uri.EscapeDataString(projectName)}/_apis/git/repositories/{Uri.EscapeDataString(repositoryId)}?api-version={ApiVersion}";
        try
        {
            return await GetJsonAsync<AdoRepository>(url, cancellationToken).ConfigureAwait(false);
        }
        catch (HttpRequestException ex) when (ex.Message.Contains(" 404 ", StringComparison.Ordinal))
        {
            return null;
        }
    }

    /// <summary>
    /// Fetch the raw text content of a file from a Git repository hosted in
    /// Azure DevOps. Returns <c>null</c> if the file is missing (404).
    /// </summary>
    public async Task<string?> GetFileContentAsync(
        string organizationName,
        string projectName,
        string repositoryId,
        string path,
        string? branch = null,
        CancellationToken cancellationToken = default)
    {
        var normalized = path.StartsWith("/", StringComparison.Ordinal) ? path : "/" + path;
        var url =
            $"https://dev.azure.com/{Uri.EscapeDataString(organizationName)}/{Uri.EscapeDataString(projectName)}/_apis/git/repositories/{Uri.EscapeDataString(repositoryId)}/items?path={Uri.EscapeDataString(normalized)}&api-version={ApiVersion}&includeContent=true&$format=text";
        if (!string.IsNullOrEmpty(branch))
        {
            url += $"&versionDescriptor.versionType=branch&versionDescriptor.version={Uri.EscapeDataString(branch!)}";
        }
        return await GetTextAsync(url, cancellationToken).ConfigureAwait(false);
    }

    /// <summary>
    /// List heads (branches) of a Git repository. Returns short branch names
    /// (without the <c>refs/heads/</c> prefix), sorted invariant-culture
    /// ordinal — same ordering as the VS Code client.
    /// </summary>
    public async Task<IReadOnlyList<string>> ListBranchesAsync(string organizationName, string projectName, string repositoryId, CancellationToken cancellationToken = default)
    {
        var url = $"https://dev.azure.com/{Uri.EscapeDataString(organizationName)}/{Uri.EscapeDataString(projectName)}/_apis/git/repositories/{Uri.EscapeDataString(repositoryId)}/refs?filter=heads/&api-version={ApiVersion}&$top=1000";
        var res = await GetJsonAsync<AdoListResponse<GitRefEntry>>(url, cancellationToken).ConfigureAwait(false);
        return res.Value
            .Select(r => r.Name.StartsWith("refs/heads/", StringComparison.Ordinal) ? r.Name.Substring("refs/heads/".Length) : r.Name)
            .OrderBy(n => n, StringComparer.Ordinal)
            .ToList();
    }

    private async Task<T> GetJsonAsync<T>(string url, CancellationToken cancellationToken)
    {
        using var request = await BuildRequestAsync(url, cancellationToken).ConfigureAwait(false);
        _logger.Debug($"GET {url}");

        using var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            await ThrowForFailureAsync(response, url, cancellationToken).ConfigureAwait(false);
        }

        var contentType = response.Content.Headers.ContentType?.MediaType ?? string.Empty;
        if (!contentType.Contains("application/json", StringComparison.OrdinalIgnoreCase))
        {
            throw new AdoUnauthorizedException(
                401,
                $"Unexpected non-JSON response from {url}. Authentication may have expired.");
        }

#if NET6_0_OR_GREATER
        await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
#else
        using var stream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
#endif
        var value = await JsonSerializer.DeserializeAsync<T>(stream, JsonOptions, cancellationToken).ConfigureAwait(false);
        return value ?? throw new InvalidOperationException($"Empty JSON response from {url}");
    }

    private async Task<string?> GetTextAsync(string url, CancellationToken cancellationToken)
    {
        using var request = await BuildRequestAsync(url, cancellationToken).ConfigureAwait(false);
        _logger.Debug($"GET {url}");

        using var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
        if (response.StatusCode == HttpStatusCode.NotFound)
        {
            return null;
        }
        if (!response.IsSuccessStatusCode)
        {
            await ThrowForFailureAsync(response, url, cancellationToken).ConfigureAwait(false);
        }

#if NET6_0_OR_GREATER
        return await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
#else
        return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
#endif
    }

    private async Task<HttpRequestMessage> BuildRequestAsync(string url, CancellationToken cancellationToken)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.Accept.ParseAdd("application/json");

        if (_authProvider is not null)
        {
            var header = await _authProvider.GetAuthHeaderAsync(cancellationToken).ConfigureAwait(false);
            if (header is not null)
            {
                request.Headers.Authorization = header;
            }
        }

        return request;
    }

    private async Task ThrowForFailureAsync(HttpResponseMessage response, string url, CancellationToken cancellationToken)
    {
        string? body = null;
        try
        {
#if NET6_0_OR_GREATER
            body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
#else
            body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
#endif
            if (body.Length > 500)
            {
                body = body.Substring(0, 500) + "…";
            }
        }
        catch
        {
            // ignored
        }

        var msg = $"ADO REST call failed: {(int)response.StatusCode} {response.ReasonPhrase} for {url}"
            + (string.IsNullOrEmpty(body) ? string.Empty : $" :: {body}");
        _logger.Error(msg);

        if (response.StatusCode == HttpStatusCode.Unauthorized || response.StatusCode == HttpStatusCode.Forbidden)
        {
            throw new AdoUnauthorizedException(
                (int)response.StatusCode,
                $"Azure DevOps rejected the credentials ({(int)response.StatusCode}). The stored token may be expired or revoked.");
        }

        throw new HttpRequestException(msg);
    }
}
